### PR TITLE
Releases with git tags (#17)

### PR DESCRIPTION
**Pull request by tom-schoener**
Issue #17

### CODE REVIEW

File: release/release.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  # Uncommitted changes
+  echo "Working directory is not clean. Please commit the changes before creating a release."
+  exit 1
+fi
+
+rootDir="$(dirname "$0")"/..
+frontendDir="${rootDir}"/frontend
+pomDir="${rootDir}"
+
+currentVersion=$(grep -oP '(?<="version": ")[^"]*' "${frontendDir}"/package.json)
```
Comment (schwerlaut) [inline release/release.sh:25]: Is this how this is done?
Can't we just use something like this:
node -pe "require('./package.json').version"

File: release/release.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+# update version (package.json)
+nextTag=$(cd "${frontendDir}" && npm version "${versionIdentifier}")
+nextVersion="${nextTag:1}"
+
+# update version (pom.xml)
+xmlstarlet ed --inplace -N a="http://maven.apache.org/POM/4.0.0" -u '/a:project/a:version' -v "${nextVersion}" "${pomDir}"/pom.xml
```
Comment (schwerlaut) [inline release/release.sh:32]: I think calling mvn version:set would be better here.
That would especially still work if we extract sub-modules.

File: release/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Releases
+    
+Releases are created with Docker. Alternatively, directly execute the `./release/release.sh` script. First, create the Docker image:
```
Comment (schwerlaut) [inline release/README.md:3]: I think this explanation is a bit confusing and starts a bit out of nowhere.
I'd propose to start this here with a small explanation on how a release of O-Neko looks like (so naming matching git tags and versions in pom.xml and packae.json).
Then there shoule be a separate section explaining the overall "How do I build the release of say version 1.5.7".
If there are alternatives to some steps then you should not continue in regular text, cause then it is unclear whether this is still part of the alternative way or not.

Comment (schwerlaut) [reply release/README.md:3]: And if your explanation covers the release-script then you totally should mention it's mandatory parameter.

File: release/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM alpine:3.11.5
```
Comment (schwerlaut) [inline release/Dockerfile:1]: I guess this is fine for the moment, but we might consider using other base images (like the stuff from distroless e.g.)

File: release/release.sh (hunk 1 above)
Comment (tom-schoener) [reply release/release.sh:25]: Yes, this would be a better solution since this script already requires Node.

**Comment by tom-schoener** [comment]
I changed the execution environment from Docker to NodeJS. Docker had a few permission problems concerning the git repository. Also, the mvn version command had to download all dependencies if you want to set the version inside the Docker container. 
Now, you simply have to run `npm i` once and to create a release: `npm run release minor` (or whatever new version you want). The Node script handles everything else. This should work on all operating systems.

File: release/package.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "release",
+  "version": "1.0.0",
+  "description": "Creates an oneko with git tags",
```
Comment (schwerlaut) [inline release/package.json:4]: Everyone want's an oneko right?

File: release/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# Releases
+
+O-Neko releases follow the [semantic versioning guidelines](https://semver.org/). A release is identified by a git tag, for example `v1.0.0`.
+A release, the frontend (`frontend/package.json`) and the backend (`pom.xml`) share the same version. Circle CI creates a new
```
Comment (schwerlaut) [inline release/README.md:4]: That sentence is slightly awry. Wouldn't is be sufficient, to just wright "Frontent (frontend/package.json) and backend (pom.xml) are released together and therefore always have the same version"?

File: release/index.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+const setAndGetNewPackageJsonVersion = (packageJsonDir, versionId) =>
+  new Promise((res, rej) => {
+    childProcess.exec(
+      `cd ${packageJsonDir} && npm version ${versionId}`,
+      (e, version) => (e ? rej(e) : res(version))
+    );
+  }).then(s => s.replace(/(\r\n|\n|\r)/gm, ""));
+
+const setPomXmlVersion = (pomXmlDir, versionId) => {
+  new Promise((res, rej) => {
```
Comment (tom-schoener) [inline release/index.js:79]: This function should return the promise.

Comment (schwerlaut) [reply release/index.js:79]: Good point.
Right now the await from line 39 will not wait for anything.

**Pull request merged**